File: go/pascals-triangle/pascals_triangle.go
package pascal

func Triangle(n int) [][]int {
	output := [][]int{}
	if n <= 0 {
		return output
	}
	output = append(output, []int{1})
	for i := 1; i < n; i++ {
		row := []int{}
		for j := 0; j <= i; j++ {
			if j == 0 || j == i {
				row = append(row, 1)
			} else {
				row = append(row, output[i-1][j-1] + output[i-1][j])
			}
		}
		output = append(output, row)
	}
	return output
}

File: elixir/pascals-triangle/lib/pascals_triangle.ex
defmodule PascalsTriangle do
  @doc """
  Calculates the rows of a pascal triangle
  with the given height
  """
  @spec rows(integer) :: [[integer]]
  def rows(num) do
    construct_rows([[1]], num)
  end

  defp construct_rows(rows, req_rows) when length(rows) == req_rows, do: Enum.reverse(rows)
  defp construct_rows(prev_rows = [last_row | _], req_rows) do
    construct_rows([create_row(last_row) | prev_rows], req_rows)
  end

  defp create_row(prev_row) do
    Enum.map(Enum.zip([[0 | prev_row], prev_row]), fn {x1, x2} -> x1 + x2 end) ++ [1]
  end
end
